graphql/src/query/ext.rs

```diff
@@ -24,3 +34,83 @@ impl ValueExt for q::Value {
         }
     }
 }
+
+pub enum BlockLocator {
+    Hash(H256),
+    Number(BlockNumber),
+}
+
+pub struct BlockConstraint {
+    pub subgraph: SubgraphDeploymentId,
+    pub block: BlockLocator,
+}
+
+pub trait FieldExt {
+    fn block_constraint<'a>(
+        &self,
+        object_type: impl Into<ObjectOrInterface<'a>>,
+    ) -> Result<Option<BlockConstraint>, QueryExecutionError>;
+}
+
+impl FieldExt for q::Field {
+    fn block_constraint<'a>(
+        &self,
+        object_type: impl Into<ObjectOrInterface<'a>>,
+    ) -> Result<Option<BlockConstraint>, QueryExecutionError> {
+        fn invalid_argument(arg: &str, field: &q::Field, value: &q::Value) -> QueryExecutionError {
+            QueryExecutionError::InvalidArgumentError(
+                field.position.clone(),
+                arg.to_owned(),
+                value.clone(),
+            )
+        }
+
+        let value =
+            self.arguments.iter().find_map(
+                |(name, value)| {
+                    if name == "block" {
+                        Some(value)
+                    } else {
+                        None
+                    }
+                },
+            );
+        if let Some(value) = value {
+            if let q::Value::Object(map) = value {
+                if map.len() != 1 || !(map.contains_key("hash") || map.contains_key("number")) {
+                    return Err(invalid_argument("block", self, value));
+                }
+                let subgraph = parse_subgraph_id(object_type)?;
+                if let Some(hash) = map.get("hash") {
+                    TryFromValue::try_from_value(hash)
+                        .map_err(|_| invalid_argument("block.hash", self, value))
+                        .map(|hash| {
+                            Some(BlockConstraint {
+                                subgraph,
+                                block: BlockLocator::Hash(hash),
+                            })
+                        })
+                } else if let Some(number_value) = map.get("number") {
+                    TryFromValue::try_from_value(number_value)
+                        .map_err(|_| invalid_argument("block.number", self, number_value))
+                        .and_then(|number: u64| {
+                            TryFrom::try_from(number)
+                                .map_err(|_| invalid_argument("block.number", self, number_value))
+                        })
+                        .map(|number| {
+                            Some(BlockConstraint {
+                                subgraph,
+                                block: BlockLocator::Number(number),
+                            })
+                        })
```

Comment:
Slightly more compact and therefor a little easier to read:

Do
```rust
let hash = map.get("hash");
let number = map.get("number");
```
first, simplify the first `if` to
```
if (map.len() != 1 || !(hash.is_some() || number.is_some()))
```
and use `match` below:
```
match (hash, number) => {
  (Some(hash), _) => { ... },
  (_, Some(number)) => { ... },
  _ => unreachable!("We already checked that there is a hash or number"),
}
```

Reply:
Fixed with commit ce5d10eb